### ISSUE TITLE
Stuck on "optimizing database" after update

### ISSUE THREAD

**Issue by alisonLFP** [signalapp/Signal-iOS#4322]
- [X] I have searched open and closed issues for duplicates
- [X] I am submitting a bug report for existing functionality that does not work as intended
- [X] This isn't a feature request or a discussion topic

----------------------------------------

### Bug description
After updating overnight, my Signal app on iOS is stuck on the "optimizing database screen" and won't load conversations. This is a different problem than #4201 because the app isn't crashing, it just hangs. 

### Steps to reproduce
- Open Signal App 
- Optimizing database screen appears and doesn't go away
- Reopening does the same thing

**Actual result:** 
App is stuck on "optimizing database" screen

**Expected result:** 
App should open and display conversations

### Device info
**Device**: iPhone 6

**iOS version**: 12.4.3

**Signal version:** I don't know, because I can't open the app settings to check. But Signal was updating before the bug happened, so I assume it's 3.0.2.1 like everyone else's iOS update?

### Link to debug log
I can't link this because I can't open the app properly.

**Comment by charlesmchen-signal** [comment]
@alisonLFP Signal iOS 3 involves a major database migration.  It may take a long time if you have a long message history, especially on older devices like your iPhone 6.  Please leave the app open for quite a while (at least 10-20 minutes) and see if that resolves this issue.  Let us know if it doesn't.

**Comment by alisonLFP** [comment]
It's working now. The database migration took about five hours.

**Comment by michaelkirk-signal** [comment]
Would you still be willing to submit a debug log so we can learn something about why the migration took so long?

**Comment by jaredlt** [comment]
As another data point, my phone has been stuck on "Optimizing Database" for 45+ minutes so far.

I have probably sent less than 10 messages using Signal ever, so it does seem like there is an issue (I should have almost no data to migrate).

Some additional info in case it helps diagnose the issue: I have had the app installed for a long time, with iOS auto-update on. I probably haven't opened the app for quite a few releases.

If/when the migration finishes I will try to submit a debug log.

Device: iPhone 7
iOS version: 13.1.3
Signal version: 3.0.6

**Comment by michaelkirk-signal** [comment]
One thing to keep in mind, you'll need to keep the app open for the migration to proceed. 

If you background the app, the OS will not let it run for long.

**Comment by jaredlt** [comment]
Thanks, I did wonder about that. I definitely had the app open in the foreground for at least an hour and it didn't finish. I haven't managed to get it to finish yet as keep forgetting to leave it on overnight. Once I do I'll submit the log.

**Comment by jaredlt** [comment]
Update to this issue: I left the migration running overnight and no luck. It ran for at least 7 hours.

After my phone alarm went off, I switched back to Signal quickly and it showed 'Optimizing Database' immediately which I took to mean iOS did not kill the migration and that it was still running (compared to the first time I opened Signal where it showed the splash screen and only after a few seconds starting showing the 'Optimizing Database'). I mention this to say that if my assumption is correct it means I ran it for over 20 hours and it did not complete (I did use my phone a few times during the day but each time switched back to Signal quickly and saw the 'Optimizing Database' immediately).

At this point I have closed the app. 

My next course of action would probably be to uninstall and reinstall the app. I barely have any messages so I am not particularly concerned about the loss, however, there may be other users in my situation who do not want to lose their data so before I do I am open to any suggestions that may help you diagnose the issue.

**Comment by michaelkirk-signal** [comment]
Sorry that this is hanging for you. Before uninstalling, would you be
willing to install a diagnostic build?

If so, please email michael@signal.org and I'll follow up.

**Comment by annarose1947** [comment]
Hi, I am having the same issue. Just updated signal on my IPhone 6 and it’s been stuck for about 20 minutes on the “optimizing database” screen. Was there ever a fix for this? If so, would love to have it. I do not want to lose my messages (although i don’t have many, it’s only 179 mb right now). Please let me know if anyone has a solution for this. I will try leaving it open overnight as well.

**Comment by jaredlt** [comment]
@annarose1947  in my case we were unable to fix it. I installed the diagnostic build and Michael determined that my app was stuck in some sort of "semi-installed" state. But we couldn’t find the root cause and I had to reinstall. 

Maybe another data point will help them find the issue.

**Comment by andrewgdotcom** [comment]
I have the same problem, have had it for months. I can't remember how long ago I first saw this issue. Like other commenters above I rarely use Signal but have had it installed for years. I have never waited long enough for the database migration to actually finish, and reading the above I now no longer want to try.

I ended up deleting and reinstalling from scratch.

**Comment by frostycellnex** [comment]
I also just experienced the behavior.  This happened after my iOS instance had auto-deleted Signal after a long lack of use, and I activated it for re-install via the "swipe-down" home screen search interface in iOS.  I ended up deleting the app and reinstalling, and that brought me through the enrollment process again, and now Signal works just fine.

**Comment by akaase** [comment]
This happened to me this morning, I have never seen this "error" before after using Signal for iOS for years now. The message "Optimizing Database" flashing on and off, with "This could take several minutes" under that. I lost my patience after a few minutes so I deleted the app and reinstalled it. If the database is being optimized, it sure would be nice to have a status bar on the screen so you can see the progress and have some idea how long it's going to take! My instinct is the app just crashed and no optimizing was happening.

**Comment by MysticalGrape** [comment]
Very bad experience. Recovery was going for over and hour and didn’t finish on iPhone 13 pro. Reinstalled and lost all the chat histories of course. Also will need to relink other devices. I had a feeling that each time the Signal went to the background (when I was using other apps) it was interrupting the process and starting again when switching back to Signal.

**Comment by akaase** [comment]
Yeah it's pretty naive software development to have a blow-off message like "This may take several minutes" and not show a exact status that stresses your users out for hours, hoping and praying the app didn't crash. If the database is fucked, just put a BSOD on the screen and tell the user to reinstall the app.

**Comment by Nateu** [comment]
Omg! Why doesn’t the transfer of an account work from ipad to iphone!?
I had the same issue, thought I could simply delete te app and reinstall. Only to discover that a fresh install required me to “transfer” my history from my ipad to iphone but it’s not letting me do that!

Just posting this here to warn ppl not to delete the app unless you wanna risk losing all history.

**Comment by MysticalGrape** [comment]
After reinstall on iPhone I had to re-link my other devices (computers and tablet), but that didn't wipe history on those linked devices. Only iPhone lost the history, but that's fine with me as I still got full history on multiple other devices. However, would be awesome to be able to cancel this database optimisation or even opt out. Still I am very thankful for Signal project and I understand that developers have limited time, so they need to keep an eye on priorities.

**Comment by miaekim** [comment]
Same issue for ios 15.3.1, signal 5.30.0

**Comment by psifertex** [comment]
Also having this same problem. I use signal once every month or two, so it's not a matter of having let it sit updating in the background many times. Been sitting at this screen for an hour now. If a debug build is helpful, happy to install it. Feel free to contact me at signal at my GH-Handle dot com.

**Comment by peterkhayes** [comment]
I also had this message hang forever. I was able to fix it by restarting my phone.

**Comment by ivan-paleo** [comment]
It happened to me right now on iPhone SE (2016) with iOS 15.4.1 and the latest version of Signal. It was stuck on this "optimization" for at least 1h. Restarting the phone fixed the problem, as suggested by @peterkhayes.

**Comment by Projanglez** [comment]
I had the same problem: Endlessly running "Optimizing Database". A restart of the phone fixed the problem. Thanks @peterkhayes for sharing your workaround.

**Comment by asininemonkey** [comment]
Can confirm the exact same issue with a phone restart required to resolve the problem.

**Comment by max-signal** [comment]
These issues should be mitigated by 80e0208bbb40a1 (included in v5.38 and higher). If anyone is still experiencing stuck migrations or migrations that take longer than ≈5 minutes, please reopen this issue.

**Comment by lorinsherry1** [comment]
hey,
I have this problem over few hours.
iPhone 8 Plus
Version OS 16.7.2

I tried to restart my iPhone, didnt help..
and removing and installing again, also didnt help.

what can I do?

**Comment by akaase** [comment]
I don't know if this issue has been resolved by the developer. I will say this hasn't happened to me in a few years. One thing I try to do is keep the database from getting too large by only keeping 90 days' worth of messages. Good luck, I hope you were able to recover the app! If not, well, it's a free app and I still love it.

________________________________
